### ISSUE TITLE
Create and use a release branch

### ISSUE THREAD

**Issue by Download** [Download/mics#17]
The abuse of the `master` branch should stop!  :)

Instead, do dependency updating, version bumping, small docs fixes etc on a separate `release` branch and only merge the tested and working stuff to `master`, so `master` will always contain the latest released code.

One question here is if/how we can make it work with Greenkeeper.

We should have gotten into the habit of doing it this way by the time we release 1.0.

**Comment by minecrawler** [comment]
Is this intended as part of GitFlow, or a custom style? Just asking for reference, so I can use the right branches for stuff.

**Comment by Download** [comment]
Well... it's the first step towards GitFlow let's say.
My idea for now is to use feature branches for code changes, then bundle them in the release branch...
Mmm come to think of it... when I use GitHub's merge it merges the feature into master! Mmmm I guess I'm going about this all wrong... :)

**Comment by Download** [comment]
Would it be ok to say that, instead of `master` only ever containing released code, it would only contain completely tested features? So it would be the tip of completed development (and as such always stable and non-breaking)?

**Comment by minecrawler** [comment]
Basically, what you explain is 90% GitFlow already, except it is missing two branch-types. Let me do a short explanation, then you can give me your opinion if we need that last step or if we should just roll with feature/release/master.

GitFlow is compromised of two "main" branches: `develop` and `main`. `develop` is forked from `main`. In order to develop features, you fork a feature-branch from `develop`, implement the feature, and merge back into `develop`. Once all the features you want in a release have been implemented and you feel it's time to do the release, you fork a release-branch from `develop`, in which you do the last changes (for example version adjustments) and tests. If everything turns out all-right, you merge the release-branch into `master`. You can then delete any feature-branch and release-branch; however, personally, I like to leave release-branches in case I need to patch and release an old version. There are additional `hotfix` branches, which can be forked from `master` and then are merged back into `master`and additionally into `develop`. For differentiation matters, the feature-branches start with `feature/` and the release-branches start with `release/`. Hotfix-branches start with `hotfix/`.

[Here](http://nvie.com/posts/a-successful-git-branching-model/) is a more verbose explanation of the whole process :)

![GitFlow Diagram](http://nvie.com/img/git-model@2x.png)

**Comment by Download** [comment]
Ok. So if I get this straight, you create a release branch *per release* right? 

And if you create a new branch based on a different branch than `master`, is GitHub smart enough to let you merge it back into that branch or does GitHub always want to merge into `master`? I never tried that actually.

I'm willing to try it out. So shall we create a `develop` branch first the next time we prepare a feature branch and then fork/merge from there?

**Comment by minecrawler** [comment]
> you create a release branch per release right?

Absolutely correct!

>  is GitHub smart enough to let you merge it back into that branch

Forking a branch and merging it somewhere are two separate things. When you have a branch you want to merge it into another branch via GitHub, you will have to create a PR. During creation, you decide from which branch into which branch you want to merge. It's those dropdowns at the top during "Compare changes" phase. The base branch (left) is the one you merge the changes into. Most of the time, though, I only create feature and hotfix branches locally, so they never even hit GitHub. Using the `--no-ff` merge option allows a better insight on the branch history later on.

> So shall we create a develop branch first the next time

Yeah :)

Apart from that, I recommend changing the default branch to `develop` and protecting the `master` branch (no force-merge + need to pass all tests + applies to administrators), which gives a little extra safety. That way, master can always be pulled as a stable version while development is going on in order in the other branches. And we can easily track the current status of development and to-be-released versions.

All that might seem like a lot of process stuff, but it makes collaboration very easy and clean. And I think we both hope for more collaboration in the future :)

**Comment by Download** [comment]
>  I recommend changing the default branch to develop and protecting the master branch (no force-merge + need to pass all tests + applies to administrators), which gives a little extra safety. 

Great idea!

> I think we both hope for more collaboration in the future :)

I do!

So let's make this all a bit easier. I invited you as collaborator on this project so you can apply these changes. That's probably a lot easier than explaining it all to me. Can you let me know if this invite gives you sufficient privilege to make these changes?
https://github.com/Download/mics/invitations

**Comment by minecrawler** [comment]
> Can you let me know if this invite gives you sufficient privilege to make these changes?

Unfortunately not. You have to set these things under Settings->Branches. Here is an example from one of my repositories:

![Branches Menu](http://i.imgur.com/cbIlt72.png)

![Branch Protection](http://i.imgur.com/vlsgKZI.png)

**Comment by Download** [comment]
Thanks for the screenshots! Very helpful. I applied the changes as shown there.  👍 

I also merged the two PRs into `release` branch.

I guess from now on we will create PRs agains `develop` and only create a new `release` branch and do version updates etc in that branch for each specific release?

I will release a new version 0.7 based on the current release branch so we can work 'GitFlow style' for all new PRs.
